### ISSUE TITLE
Explain & rectify nondeterministic failures of mempool_nu_activation.py

### ISSUE THREAD

**Issue by nuttycom** [zcash/zcash#5718]
It seems like a surprisingly large number of our nondeterministic RPC failures lately have been from mempool_nu_activation.py - does anyone have ideas as to why this test would be so vulnerable to error? One thing I noticed when looking at the code is that the way that the timeout is handled in https://github.com/zcash/zcash/blob/master/qa/rpc-tests/test_framework/util.py#L141 seems potentially problematic. If it takes a long time in the initial loop waiting for the mempool contents to converge, there may be very little of the timeout remaining in which the second loop that awaits fullyNotified status can complete.

**Comment by LarryRuane** [comment]
Here's a possible fix and also some debug code that @nuttycom had in #5719 branch while investigating this. I'm removing this code from that PR but don't want to lose it, so I'll paste it here in case it's valuable in helping investigate this:
```
diff --git a/qa/rpc-tests/mempool_nu_activation.py b/qa/rpc-tests/mempool_nu_activation.py
index 14d5df2c2..a9ee6f0b7 100755
--- a/qa/rpc-tests/mempool_nu_activation.py
+++ b/qa/rpc-tests/mempool_nu_activation.py
@@ -178,6 +178,7 @@ class MempoolUpgradeActivationTest(BitcoinTestFramework):
 
             # Mine blocks on node 1, so that the Y transactions in its mempool
             # will be cleared.
+            self.sync_all()
             self.nodes[1].generate(6)
             self.sync_all()
 
diff --git a/qa/rpc-tests/test_framework/util.py b/qa/rpc-tests/test_framework/util.py
index 4c26ce0e6..1f85a3615 100644
--- a/qa/rpc-tests/test_framework/util.py
+++ b/qa/rpc-tests/test_framework/util.py
@@ -151,15 +151,25 @@ def sync_mempools(rpc_connections, wait=0.5, timeout=60):
                 num_match = num_match+1
         if num_match == len(rpc_connections):
             break
+        print("timeout: ", str(timeout), str(wait))
         time.sleep(wait)
         timeout -= wait
 
     # Now that the mempools are in sync, wait for the internal
     # notifications to finish
     while timeout > 0:
+        if timeout % 10 == 0: 
+            print("##############################################")
+            for x in rpc_connections:
+                print(str(x.getmempoolinfo()))
         notified = [ x.getmempoolinfo()['fullyNotified'] for x in rpc_connections ]
         if notified == [ True ] * len(notified):
             return True
+        else:
+            print("----------------------------------------------")
+            for x in rpc_connections:
+                print(str(x.getmempoolinfo()))
+        print("timeout: ", str(timeout), str(wait))
         time.sleep(wait)
         timeout -= wait
 
```

**Comment by nuttycom** [comment]
Fixed by #5900
